### ISSUE TITLE
Cannot search for user attributes that contain whitespace

### ISSUE THREAD

**Issue by koplas** [keycloak/keycloak#33094]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

dist/quarkus

### Describe the bug

If you add a custom user attribute and set the value to "hello world" it is impossible to search for this attribute, as the value contains the space character. If the value is set to "helloworld" it will find the value.

This bug is not visible while searching "First name", "Last name", or "E-Mail", because custom attribute search will require an exact match to find the results.

### Version

25.0.5

### Regression

- [ ] The issue is a regression

### Expected behavior

Search for "hello world" and get a search result.

### Actual behavior

No search results are returned.

### How to Reproduce?

- Go to `Realm Settings` -> `User profile`
- Create a new attribute called `test`

- Go to `Users` and click on a user
- Go to the `test` attribute field and enter "hello world"

- Return to the user page and select `Attribute search` in the `Search user` field.
- Search for the attribute `test` with the value "hello world"
- Get zero results

- Set the `test` attribute value to "helloworld"
- Search again for `test` with the value "helloworld"
- Get as expected one result

### Anything else?

~~The bug is in `SearchQueryUtils.getFields(String query)`. The regex used in this function ignores all characters after a whitespace. The front end correctly encodes the space character as `+`.~~

Update: This is a bug in the frontend; if you search for "hello world" with the quotes, it will find the result. If you search without it will find no results. All searches that contain whitespace should be escaped with `"`.

**Comment by btirth** [comment]
Hi @jonkoops can I take this one?

**Comment by jonkoops** [comment]
Sure, I'll assign this to you.

**Comment by btirth** [comment]
Hi @jonkoops, should I handle this in frontend (as mentioned in Update part of issue description) or backend?

**Comment by jonkoops** [comment]
Sure, assigned :+1:

**Comment by jonkoops** [comment]
I'd say this should be solved on the REST API level so that anyone using the API benefits.

**Comment by codespearhead** [comment]
> [!NOTE]  
TL;DR: I plan to make a PR. It would be great if you could assign this issue to me as well. @btirth, let me know if you're still interested in helping to get it fixed.

This issue was partially and indirectly fixed by commit [1] from PR [2], which is functionally the same [3][4][5] as the workaround described in this issue: enclosing the attribute value being queried in double quotes.

I confirm that in version “26.0.0”, you no longer need the double quotes when searching for a single tag. However, if you add another attribute to the search, the problem still occurs.

I would argue that the root cause of this issue is the same as what was reported in issue [6][7]. Specifically, it relates to the need to change “filter.value” to “encodeURIComponent(filter.value)” here [8]. The problem arises because the value key of the attributes object is not URI-encoded. As a result, anything after the first whitespace in that key is ignored, since whitespace is used as a delimiter to separate the attributes. You can observe this in the “q” parameter of the request sent to the “/ui-ext/brute-force-user” endpoint when pressing the search button, which shows that the issue is frontend-related.

[1] a339e79d3ecbd35c12abd3c67717fc5ea466b415 (title: added exact search option to attributes)
[2] https://github.com/keycloak/keycloak/pull/33465
[3] https://github.com/keycloak/keycloak/blob/14f14152dee691d61274fa73638fc67971243be9/js/apps/admin-ui/src/components/users/UserDataTable.tsx#L47-L50
[4] https://github.com/keycloak/keycloak/blob/14f14152dee691d61274fa73638fc67971243be9/js/apps/admin-ui/src/components/users/UserDataTable.tsx#L237-L241
[5] https://github.com/keycloak/keycloak/blob/14f14152dee691d61274fa73638fc67971243be9/js/apps/admin-ui/src/components/users/UserDataTable.tsx#L243-L247
[6] https://github.com/keycloak/keycloak/issues/33122
[7] https://github.com/keycloak/keycloak/issues/33122#issuecomment-2392135977
[8] https://github.com/keycloak/keycloak/blob/6e60287290cb406934646ccf95ae68cf6440a310/js/apps/admin-ui/src/components/users/UserDataTable.tsx#L239

***
## How to Reproduce
***

01. Log in to the Keycloak Admin Console.
02. On the left-hand side, select “Realm settings”.
03. Click the “User profile” tab.
04. Click “Create attribute”.
05. Add “att1” to the “Attribute [Name]” field, then click “Create”.
06. On the left-hand side, select "Users".
07. Select the “User list” tab, then click "Add user".
08. In the "Create user" form, add the information from the first row in the table below to the 
“Create user” form to create “user1”, then click "Create". Repeat steps 07 through 08 for the second row to create “user2” and “user3”.

| Name  | att1      | att2 |
|-|-|-|
| user1 | a b | |
| user2 | a_b | |
| user3 | | c d |

10. On the left-hand side, select "Users".
11. Select the “User list” tab, then click “Default search” and choose “Attribute search”.
12. Click “Search attributes”.
13. Add the information below to the “Select attributes” form, click the checkmark to add the attribute to the filter, then click “Search.”

# TODO: fill in the truth table, then turn it into tests
## The table below, as well as the previous one, may need some adjustments

| Test Number | Test Status | Value[Key=att1] | Value[Key=att2] | Current        | Expected   |
|-------------|-------------|-----------------|-----------------|----------------|------------|
| 01          | fail        | a b             | [empty]               | user1, user2   |   user1         |
| 02          |         | "a b"           | [empty]               |           |            |
| 03          |             | a_b             | [empty]               |   |            |
| 04          |             | "a_b"           | [empty]               |    |            |
| 05          |             | a b             | c d             |                |            |
| 06          |             | "a b"           | c d             |                |            |
| 07          |             | a_b             | c d             |                |            |
| 08          |             | "a_b"           | c d             |                |            |
| 09          |             | a b             | "c d"           |                |            |
| 10          |             | "a b"           | "c d"           |                |            |
| 11          |             | a_b             | "c d"           |                |            |
| 12          |             | "a_b"           | "c d"           |                |            |
| 13          |             | a b             | c_d             |                |            |
| 14          |             | "a b"           | c_d             |                |            |
| 15          |             | a_b             | c_d             |                |            |
| 16          |             | "a_b"           | c_d             |                |            |
| 17          |             | a b             | "c_d"           |                |            |
| 18          |             | "a b"           | "c_d"           |                |            |
| 19          |             | a_b             | "c_d"           |                |            |
| 20          |             | "a_b"           | "c_d"           |                |            |

**Comment by Amer-Jabar** [comment]
In case you didn't see the updates, you can search for values with whitespace. You just need to escape it with ".
